core-bundle/src/Util/SimpleTokenParser.php

```diff
@@ -137,25 +140,39 @@ private function canUseExpressionLanguage(array $data): bool
 
     private function evaluateExpression(string $expression, array $data, bool $canUseExpressionLanguage): bool
     {
-        if ($canUseExpressionLanguage) {
-            try {
-                return $this->expressionLanguage->evaluate($expression, $data);
-            } catch (\Exception $e) {
-                throw new \InvalidArgumentException($e->getMessage(), 0, $e);
-            }
+        if (!$canUseExpressionLanguage) {
+            return $this->evaluateExpressionLegacy($expression, $data);
+        }
+
+        $unmatchedVariables = array_diff($this->getVariables($expression), array_keys($data));
+
+        if (!empty($unmatchedVariables)) {
+            $this->logUnmatchedVariables(...$unmatchedVariables);
+
+            // Define variables that weren't provided with the value 'null'
+            $data = array_merge(
+                array_combine($unmatchedVariables, array_fill(0, \count($unmatchedVariables), null)),
+                $data
+            );
         }
 
-        // Legacy code
+        try {
+            return (bool) $this->expressionLanguage->evaluate($expression, $data);
+        } catch (SyntaxError $e) {
+            throw new \InvalidArgumentException($e->getMessage(), 0, $e);
+        }
+    }
+
+    private function evaluateExpressionLegacy(string $expression, array $data): bool
+    {
         if (!preg_match('/^([^=!<>\s]+) *([=!<>]+)(.+)$/s', $expression, $matches)) {
             return false;
         }
 
         [, $token, $operator, $value] = $matches;
 
         if (!\array_key_exists($token, $data)) {
-            if (null !== $this->logger) {
-                $this->logger->log(LogLevel::INFO, sprintf('Tried to evaluate unknown simple token "%s".', $token));
-            }
+            $this->logUnmatchedVariables($token);
 
             return false;
```

Comment:
We need to fix it here too, right?

Reply:
Done in d7995b689f23d4520289fd1c046293ac3b356b3b.